contracts/ThreeHeapOrdering.sol

```diff
@@ -0,0 +1,306 @@
+            else if (formerValue < newValue) increase(_heap, _id, newValue, _maxSortedUsers);
+            else decrease(_heap, _id, newValue);
+        }
+    }
+
+    /// PRIVATE ///
+
+    /// @notice Computes a new suitable size from `_size` that is smaller than `_maxSortedUsers`.
+    /// @dev We use division by 2 to remove the leaves of the heap.
+    /// @param _size The old size of the heap.
+    /// @param _maxSortedUsers The maximum size of the heap.
+    /// @return The new size computed.
+    function computeSize(uint256 _size, uint256 _maxSortedUsers) private pure returns (uint256) {
+        while (_size >= _maxSortedUsers) _size /= 3;
+        return _size;
+    }
+
+    /// @notice Sets `_index` in the `_heap` to be `_account`.
+    /// @dev The heap may lose its invariant about the order of the values stored.
+    /// @dev Only call this function with an index within array's bounds.
+    /// @param _heap The heap to modify.
+    /// @param _index The index of the account in the heap to be set.
+    /// @param _account The account to set the `_index` to.
+    function setAccount(
+        HeapArray storage _heap,
+        uint256 _index,
+        Account memory _account
+    ) private {
+        _heap.accounts[_index] = _account;
+        _heap.indexes[_account.id] = _index;
+    }
+
+    /// @notice Moves an account up the heap until its value is smaller than the one of its parent.
+    /// @dev This functions restores the invariant about the order of the values stored when the account to move is the only one with value greater than what it should be.
+    /// @param _heap The heap to modify.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftUp(
+        HeapArray storage _heap,
+        Account memory _accountToShift,
+        uint256 _index
+    ) private {
+        Account memory parentAccount;
+        uint256 parentIndex;
+        while (
+            _index > ROOT &&
+            _accountToShift.value >
+            (parentAccount = _heap.accounts[parentIndex = (_index - 1) / 3]).value
+        ) {
+            setAccount(_heap, _index, parentAccount);
+            _index = parentIndex;
+        }
+        setAccount(_heap, _index, _accountToShift);
+    }
+
+    /// @notice Moves an account down the heap until its value is greater than the ones of its children.
+    /// @dev This functions restores the invariant about the order of the values stored when the account to move is the only one with value smaller than what it should be.
+    /// @param _heap The heap to modify.
+    /// @param _accountToShift The account to move.
+    /// @param _index The index of the account to move.
+    function shiftDown(
+        HeapArray storage _heap,
+        Account memory _accountToShift,
+        uint256 _index
+    ) private {
+        uint256 size = _heap.size;
+        Account memory targetAccount = _accountToShift;
```

Comment:
There are 2 main changes that require changing the naming:
- we use a while loop to go through the three child. This was changed from the MEP solution because it greatly reduces the code size
- we pass the account to shift as an argument. It is a small optimization not in the HeapOrdering.